// cron-parser/dist/fields/CronMonth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CronMonth = void 0;
const CronField_1 = require("./CronField");
const MIN_MONTH = 1;
const MAX_MONTH = 12;
const MONTH_CHARS = Object.freeze([]);
const DAYS_IN_MONTH = Object.freeze([31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]);
/**
 * Represents the "day of the month" field within a cron expression.
 * @class CronDayOfMonth
 * @extends CronField
 */
class CronMonth extends CronField_1.CronField {
    static get min() {
        return MIN_MONTH;
    }
    static get max() {
        return MAX_MONTH;
    }
    static get chars() {
        return MONTH_CHARS;
    }
    static get daysInMonth() {
        return DAYS_IN_MONTH;
    }
    /**
     * CronDayOfMonth constructor. Initializes the "day of the month" field with the provided values.
     * @param {MonthRange[]} values - Values for the "day of the month" field
     * @param {boolean} [wildcard=false] - Whether this field is a wildcard
     */
    constructor(values, wildcard = false) {
        super(values, wildcard);
        this.validate();
    }
    /**
     * Returns an array of allowed values for the "day of the month" field.
     * @returns {MonthRange[]}
     */
    get values() {
        return super.values;
    }
}
exports.CronMonth = CronMonth;
